File: model_chromatogram/methods/pressure.py
import os
from model_chromatogram.user_parameters import JULIA_PARAMTERS

if JULIA_PARAMTERS["julia"] is not None:
    os.environ["PYTHON_JULIAPKG_EXE"] = JULIA_PARAMTERS["julia"]

    from juliacall import Main as jl

    jl.include("./model_chromatogram/methods/pressure.jl")
    use_julia_fit = True
else:
    use_julia_fit = False


from model_chromatogram.system import System
import pandas as pd
import numpy as np
from model_chromatogram.methods.viscosity import (
    MethanolViscosity,
    AcetonitrileViscosity,
    TetrahydrofuranViscosity,
)
from typing import Union


class PressureDriver:
    initial_pressure_guess = 200  # in bar

    meoh_viscosity = MethanolViscosity()
    acn_viscosity = AcetonitrileViscosity()
    thf_viscosity = TetrahydrofuranViscosity()

    def __init__(
        self,
        system: System,
        solvent_profile: pd.DataFrame,
        column_permeability_factor=20,
    ) -> None:
        self.column_length = system.column.length_mm * 1e-3  # in m
        self.particle_diameter = system.column.particle_diameter_um * 1e-6  # in m
        self.particle_sphericity = 0.95
        self.porosity = system.column.porosity

        self.solvent_profile = solvent_profile

        # add incremental volume to the profile table
        time_delta = self.solvent_profile["time"][1] - self.solvent_profile["time"][0]
        self.solvent_profile["incremental_CV"] = (
            self.solvent_profile["flow"] * time_delta / system.column.volume
        )
        self.solvent_profile["incremental_length"] = (
            self.solvent_profile["incremental_CV"] * self.column_length
        )
        # flow in cm**3/min, column inner_diameter in mm. Convert to m/s
        self.solvent_profile["superficial_velocity"] = (
            self.solvent_profile["flow"]
            / (system.column.id_mm**2 * np.pi / 400)
            / (100 * 60)
        )
        self.solvent_profile["meoh_x"] = (
            self.meoh_viscosity.compute_molar_fraction_from_percent(
                self.solvent_profile["percent_meoh"]
            )
        )
        self.solvent_profile["acn_x"] = (
            self.thf_viscosity.compute_molar_fraction_from_percent(
                self.solvent_profile["percent_acn"]
            )
        )
        self.solvent_profile["thf_x"] = (
            self.thf_viscosity.compute_molar_fraction_from_percent(
                self.solvent_profile["percent_thf"]
            )
        )

        self.kozeny_carman_multiplier = (
            column_permeability_factor
            * 150
            / (self.particle_diameter * self.particle_sphericity) ** 2
            * (1 - self.porosity) ** 2
            / (self.porosity**3)
        )

    def kozeny_carman_pressure(
        self,
        viscosity: Union[float, np.array],
        velocity: float,
        length: Union[float, np.array],
    ) -> float:
        """Calculate the Kozeny-Carman factor.

        Args:
            viscosity (float): The viscosity of the mobile phase (in cP).
            velocity (float): The velocity of the mobile phase (in cm/min).
            length (float): The length of the section (in mm).
        Returns:
            float: The pressure drop across the section (in bar).
        """
        # Kozeny–Carman equation for pressure drop
        pressure = self.kozeny_carman_multiplier * viscosity * velocity * length

        return pressure * 1e-8  # convert to bar

    def total_viscosity(self, meoh_x, acn_x, thf_x, temp, pressure) -> float:
        """Calculate the total viscosity of the mobile phase.

        Args:
            meoh_x (float): The molar fraction of methanol.
            acn_x (float): The molar fraction of acetonitrile.
            thf_x (float): The molar fraction of tetrahydrofuran.
            temp (float): The temperature of the mobile phase (in Celsius).
            pressure (float): The pressure of the mobile phase (in bar).

        Returns:
            float: The viscosity of the mobile phase (in cP).
        """
        meoh_nu = self.meoh_viscosity.interpolate_viscosity(pressure, temp, meoh_x)

        if acn_x == 0:
            acn_nu = 1
        else:
            acn_nu = self.acn_viscosity.interpolate_viscosity(pressure, temp, acn_x)
        if thf_x == 0:
            thf_nu = 1
        else:
            thf_nu = self.thf_viscosity.interpolate_viscosity(pressure, temp, thf_x)

        if meoh_nu == thf_nu == acn_nu == 1:
            return self.meoh_viscosity.interpolate_viscosity(pressure, temp, meoh_x)

        else:
            water_meoh_x = 1 - (acn_x + thf_x)
            return np.exp(
                np.log(meoh_nu) * water_meoh_x
                + np.log(acn_nu) * acn_x
                + np.log(thf_nu) * thf_x
            )

    def calculate_pressure_simple(self, tol=1e-5) -> None:
        """Calculate the pressure at each time point, treating the solvent composition as uniform across the column;
        this is an approximate method.

        Args:
            tol (float, optional): The tolerance for the pressure calculation. Defaults to 1e-6.

        Returns:
            np.array(float): The pressure of the system at each time point.
        """
        temp = self.solvent_profile["temperature"].to_numpy()
        v = self.solvent_profile["superficial_velocity"].to_numpy()
        meoh_x = self.solvent_profile["meoh_x"].to_numpy()
        acn_x = self.solvent_profile["acn_x"].to_numpy()
        thf_x = self.solvent_profile["thf_x"].to_numpy()

        pressure_guess = self.initial_pressure_guess
        pressures = np.zeros(len(temp))
        for i in range(len(temp)):
            delta = np.inf
            while delta > tol:
                viscosity_guess = self.total_viscosity(
                    meoh_x[i], acn_x[i], thf_x[i], temp[i], pressure_guess
                )
                pressure = self.kozeny_carman_pressure(
                    viscosity_guess, v[i], self.column_length
                )
                delta = np.abs(pressure_guess - pressure)
                pressure_guess = pressure

            pressures[i] = pressure_guess

        return pressures

    def calculate_pressure_finite_difference(
        self, tol=1e-5, recalculate_viscosities=False
    ) -> None:
        """Calculate the pressure at each time point using a finite difference method.

        Args:
            tol (float, optional): The tolerance for the pressure calculation. Defaults to 1e-6.

        Returns:
            np.array(float): The pressure of the system at each time point.
        """
        temp = self.solvent_profile["temperature"].to_numpy()
        v = self.solvent_profile["superficial_velocity"].to_numpy()
        meoh_x = self.solvent_profile["meoh_x"].to_numpy()
        acn_x = self.solvent_profile["acn_x"].to_numpy()
        thf_x = self.solvent_profile["thf_x"].to_numpy()
        incremental_CV = self.solvent_profile["incremental_CV"].to_numpy()
        incremental_lengths = self.solvent_profile["incremental_length"].to_numpy()

        pressure_guess = self.initial_pressure_guess
        pressures = np.zeros(len(temp))
        viscosities = np.zeros(len(temp))
        for i in range(len(temp)):
            delta = np.inf

            column_fraction_accounted_for = 0
            lower_index = i
            column_unaccounted_for = 0
            column_overaccounted_for = 0
            while column_fraction_accounted_for < 1:
                column_fraction_accounted_for += incremental_CV[lower_index]
                if column_fraction_accounted_for >= 1:
                    column_overaccounted_for = column_fraction_accounted_for - 1
                    break
                if lower_index == 0:
                    column_unaccounted_for = 1 - column_fraction_accounted_for
                    break
                lower_index -= 1

            upper_index = i + 1

            while delta > tol:
                if recalculate_viscosities:
                    # viscosities for each segment
                    curr_viscosities = self.total_viscosity(
                        meoh_x[lower_index:upper_index],
                        acn_x[lower_index:upper_index],
                        thf_x[lower_index:upper_index],
                        temp[lower_index:upper_index],
                        pressure_guess,
                    )
                else:
                    viscosities[i] = self.total_viscosity(
                        meoh_x[i],
                        acn_x[i],
                        thf_x[i],
                        temp[i],
                        pressure_guess,
                    )
                    curr_viscosities = viscosities[lower_index:upper_index].copy()

                # incremental pressures for each segment
                incremental_pressures = self.kozeny_carman_pressure(
                    curr_viscosities,
                    v[lower_index:upper_index],
                    incremental_lengths[lower_index:upper_index],
                )

                pressure = np.sum(incremental_pressures)
                if column_unaccounted_for > 0:
                    pressure += (
                        column_unaccounted_for
                        / (incremental_CV[lower_index])
                        * incremental_pressures[0]
                    )
                elif column_overaccounted_for > 0:
                    pressure -= (
                        column_overaccounted_for
                        / (incremental_CV[lower_index])
                        * incremental_pressures[-1]
                    )

                delta = np.abs(pressure_guess - pressure)
                pressure_guess = pressure

            pressures[i] = pressure_guess
        return pressures


def calculate_pressure(
    mobile_phase: dict, profile_table: pd.DataFrame, system: System
) -> float:
    """Calculate the pressure of the system.

    Args:
        mobile_phase (dict): The mobile phase dictionary.
        profile_table (dict): The profile table dictionary.
        system (System): The system object.

    Returns:
        np.array(float): The pressure of the system at each time point.
    """

    # first calculate which solvents are used in the mobile phase.
    # first calculate which solvents are used in the mobile phase.

    profile_table_copy = profile_table.copy(deep=True)

    for solvent in mobile_phase:
        old_name = f"percent_{solvent['id'].lower()}"
        name = solvent["name"].lower()
        if name in ["water", "h2o", "buffer"]:
            new_name = "percent_water"
        elif name in ["acetonitrile", "acn"]:
            new_name = "percent_acn"
        elif name in ["methanol", "meoh"]:
            new_name = "percent_meoh"
        elif name in ["thf", "tetrahydrofuran"]:
            new_name = "percent_thf"
        else:
            raise ValueError(f"Unknown solvent: {solvent["name"]}")
        profile_table_copy.rename(columns={old_name: new_name}, inplace=True)

    col_struct = [
        system.column.length_mm * 1e-3,
        system.column.particle_diameter_um * 1e-6,
        0.95,
        system.column.porosity,
        system.column.volume,
        system.column.id_mm,
    ]
    if use_julia_fit:
        pressure = jl.Pressure.pressure_driver(profile_table_copy, col_struct)

    else:
        pressure_driver = PressureDriver(system, profile_table_copy)
        pressure = pressure_driver.calculate_pressure_finite_difference()
    return pressure

File: model_chromatogram/methods/pressure.jl
module Pressure

export pressure_driver

using DataFrames
include("./viscosity.jl")
using .ViscosityModel

METHANOL = Dict{String,Float64}(
    "density" => 0.792,  # g/cm^3
    "mw" => 32.04,  # g/mol
)
WATER = Dict{String,Float64}(
    "density" => 0.997,  # g/cm^3
    "mw" => 18.01528,  # g/mol
)
ACETONITRILE = Dict{String,Float64}(
    "density" => 0.786,  # g/cm^3
    "mw" => 41.05,  # g/mol
)
THF = Dict{String,Float64}(
    "density" => 0.886,  # g/cm^3
    "mw" => 72.11,  # g/mol
)



struct Column
    length::Float64
    particle_diameter::Float64
    particle_sphericity::Float64
    porosity::Float64
    volume::Float64
    inner_diameter::Float64
end


function calculate_molar_fraction(org_type::String, org_percent::Vector{Float64})::Vector{Float64}
    if org_type == "methanol"
        org_dict = METHANOL
    elseif org_type == "acetonitrile"
        org_dict = ACETONITRILE
    elseif org_type == "thf"
        org_dict = THF
    else
        throw(ArgumentError("Unknown organic type: $org_type"))
    end
    a::Vector{Float64} = org_percent .* org_dict["density"] .* org_dict["mw"]
    b::Vector{Float64} = (100 .- org_percent) .* WATER["density"] .* WATER["mw"]
    return a ./ (a .+ b)
end


function update_solvent_profile!(sp_df::DataFrames.DataFrame, column::Column)
    time_delta::Float64 = sp_df.time[2] - sp_df.time[1]
    sp_df.incremental_CV = sp_df.flow .* time_delta ./ column.volume
    sp_df.incremental_length = sp_df.incremental_CV .* column.length
    sp_df.superficial_velocity = sp_df.flow ./ (column.inner_diameter^2 * π / 400) ./ (100 * 60)
    sp_df.meoh_x = calculate_molar_fraction("methanol", sp_df.percent_meoh)
    sp_df.acn_x = calculate_molar_fraction("acetonitrile", sp_df.percent_acn)
    sp_df.thf_x = calculate_molar_fraction("thf", sp_df.percent_thf)
end

function total_viscosity(
    meoh_x::Vector{Float64},
    acn_x::Vector{Float64},
    thf_x::Vector{Float64},
    t::Vector{Float64},
    p::Float64)::Vector{Float64}

    threshold = 1e-6
    if acn_x + meoh_x + thf_x < threshold
        return viscosity("meoh", t, meoh_x, p)
    end
    water_meoh_x = 1.0 .- (acn_x .+ thf_x)
    meoh_viscosity = viscosity("meoh", t, meoh_x, p)
    acn_viscosity = acn_x > threshold ? viscosity("acn", t, acn_x, p) : 1.0
    thf_viscosity = thf_x > threshold ? viscosity("thf", t, thf_x, p) : 1.0
    return exp.(water_meoh_x .* log.(meoh_viscosity) .+ acn_x .* log.(acn_viscosity) .+ thf_x .* log.(thf_viscosity))
end #total_viscosity

function total_viscosity(
    meoh_x::Float64,
    acn_x::Float64,
    thf_x::Float64,
    t::Float64,
    p::Float64)::Float64

    threshold = 1e-6
    if typeof(meoh_x) == Float64 && acn_x + meoh_x + thf_x < threshold
        return viscosity("meoh", t, meoh_x, p)
    end
    water_meoh_x = 1.0 .- (acn_x .+ thf_x)
    meoh_viscosity = viscosity("meoh", t, meoh_x, p)
    acn_viscosity = acn_x > threshold ? viscosity("acn", t, acn_x, p) : 1.0
    thf_viscosity = thf_x > threshold ? viscosity("thf", t, thf_x, p) : 1.0
    return exp(water_meoh_x * log(meoh_viscosity) + acn_x * log(acn_viscosity) + thf_x * log(thf_viscosity))
end #total_viscosity

function kozeny_carman_model(
    kozeny_carman::Float64,
    v::Vector{Float64},
    η::Vector{Float64},
    l::Vector{Float64})::Vector{Float64}
    pressure = kozeny_carman .* η .* l .* v
    return pressure .* 1e-8
end #kozeny_carman_model

function kozeny_carman_model(
    kozeny_carman::Float64,
    v::Float64,
    η::Float64,
    l::Float64)::Float64
    pressure = kozeny_carman .* η .* l .* v
    return pressure .* 1e-8
end #kozeny_carman_model

function pressure_driver(
    solvent_profile,
    column::AbstractArray;
    column_permeability_factor::Float64=20.0,
    tolerance::Float64=1e-5,
    recalculate_viscosities::Bool=false,
    initial_pressure_guess::Float64=200.0)::Vector{Float64}

    column = Column(column...)
    sp_df::DataFrames.DataFrame = solvent_profile |> DataFrames.DataFrame
    kozeny_carman::Float64 = column_permeability_factor * 150 / (column.particle_diameter * column.particle_sphericity)^2 * (1 - column.porosity)^2 / column.porosity^3
    update_solvent_profile!(sp_df, column)

    viscosities::Vector{Float64} = zeros(length(sp_df.time))
    pressures::Vector{Float64} = zeros(length(sp_df.time))

    for i in 1:length(sp_df.time)
        δ::Float64 = Inf64
        column_fraction_accounted_for::Float64 = 0.0
        lower_index::Int64 = i
        column_unaccounted_for::Float64 = 0
        column_overaccounted_for::Float64 = 0
        while column_fraction_accounted_for < 1
            column_fraction_accounted_for += sp_df.incremental_CV[lower_index]
            if column_fraction_accounted_for >= 1
                column_overaccounted_for = column_fraction_accounted_for - 1
                break
            end
            if lower_index == 1
                column_unaccounted_for = 1 - column_fraction_accounted_for
                break
            end
            lower_index -= 1
        end

        pressure::Float64 = 0.0
        while δ > tolerance
            if recalculate_viscosities
                meoh_x = sp_df.meoh_x[lower_index:i]
                acn_x = sp_df.acn_x[lower_index:i]
                thf_x = sp_df.thf_x[lower_index:i]
                t = sp_df.temperature[lower_index:i]
                p = initial_pressure_guess
                current_η = total_viscosity(meoh_x, acn_x, thf_x, t, p)
            else
                η = total_viscosity(sp_df.meoh_x[i], sp_df.acn_x[i], sp_df.thf_x[i], sp_df.temperature[i], initial_pressure_guess)

                viscosities[i] = η
                current_η = viscosities[lower_index:i]
            end

            incremental_pressures = kozeny_carman_model(kozeny_carman, sp_df.superficial_velocity[lower_index:i], current_η, sp_df.incremental_length[lower_index:i])

            pressure = sum(incremental_pressures)
            if column_unaccounted_for > 0
                pressure += column_unaccounted_for / sp_df.incremental_CV[lower_index] * incremental_pressures[1]
            elseif column_overaccounted_for > 0
                pressure -= column_overaccounted_for / sp_df.incremental_CV[lower_index] * incremental_pressures[end]
            end

            δ = abs(pressure - initial_pressure_guess)
            initial_pressure_guess = pressure
        end #while δ > tolerance
        pressures[i] = pressure

    end

    return pressures

end #pressure_driver

end #module